File: Game Project (Java in Eclipse)/src/com/Foofles/Mover.java
package com.Foofles;

public class Mover extends DynamicObject {
	public byte uFunctionType;
	public byte uMoverType;
	public float fSpeed;
	public float Width;
	public float Pause;
	public float OriginX;
	public float OriginY;
	public float Sync;  //Offset of Time
	public float Mod;
	
	public static final byte FunctionSin = 1;
	public static final byte FunctionLin = 2;
	public static final byte HorizontalMover = 0;
	public static final byte VerticalMover = 1;
	public static final byte Fader = 2;
	public static final byte Slammer = 3;
	

	@Override
	public boolean Update(){
		float DeltaX = 0.0f;
		float DeltaY = 0.0f;
		if(uMoverType == Slammer){
			Mod = FoofMath.SineWave(Globals.runtimeSeconds + Sync, fSpeed, Width, Pause, Pause);
			final float NewY = Mod + OriginY;
			DeltaY = NewY - Y;
			Y = Mod + OriginY;
			X = OriginX;
		}else{
			Mod = FoofMath.TriangleWave(Globals.runtimeSeconds + Sync, fSpeed, Width, Pause, 0);
			//final float Mod = (float) Math.sin(Globals.runtimeSeconds);
			switch(uMoverType){
				case HorizontalMover:{
					final float NewX = Mod + OriginX;
					DeltaX = NewX - X;
					X = NewX;
					Y = OriginY;
					break;
				}
				case VerticalMover:{
					final float NewY = Mod + OriginY;
					DeltaY = NewY - Y;
					Y = Mod + OriginY;
					X = OriginX;
					break;
				}
				default:
					break;
			}
		}
		//if(DeltaX != 0.0f || DeltaY != 0.0f){
		VelocityX = DeltaX;
		CD.Translate(DeltaX, DeltaY);
		return true;
	}
	
	@Override
	public void ComputeVisibleArea(){
		final float WidthRight = CD.Right - X;
		final float WidthLeft = X - CD.Left;

		if(uMoverType == Mover.VerticalMover || uMoverType == Mover.Fader){
			VisibilityMax = OriginX + WidthRight;
			VisibilityMin = OriginX - WidthLeft;
		}else{
			final float DoubleWidth = Width * 2;
			VisibilityMax = OriginX + DoubleWidth + WidthRight;
			VisibilityMin = OriginX - DoubleWidth - WidthLeft;
		}
		return;
	}
	@Override
	void Render(final float CamX, final float CamY, final SpriteBatch mSpriteBatch){
		final int SpriteID = (mAnimation.bIsPlaying) ? mAnimation.AdvanceFrame() : Info.nSpriteID;
		if(uMoverType == Mover.Fader){
			mAnimation.bIsPlaying = true;
			if(Mod > 0.0f){
				Globals.mGameGraph.SetColor(1.0f, 1.0f, 1.0f, (float)Math.sin(Mod)*2.0f);
				mSpriteBatch.AddSprite(X - CamX, Y - CamY, SpriteID, Info.fDefaultScale, bFlip, Globals.mGameGraph.Color);
			}
		}else{
			mSpriteBatch.AddSprite(X - CamX, Y - CamY, SpriteID, Info.fDefaultScale, bFlip, null);
		}
		return;
	}
}

File: FoofLevelEditor (C++)/FoofLevelEditor/Mover.cpp
#include "Mover.h"
#include "Globals.h"

Mover::Mover(void){
	bIsSelected = false;
	uFunctionType = Mover::FunctionLin;
	uMoverType = Mover::HorizonalMover;
	fSpeed = 0.55f;
	Width = 0.25f;
	Sync = 0.0f;
	Pause = 0.0f;
}
Mover::~Mover(void){
}
void Mover::Duplicate(PropStruct ** ppPropStruct){
	*ppPropStruct = new Mover;
	memcpy(*ppPropStruct, this, sizeof(Mover));
}
void Mover::Update(){
		if(uMoverType == Slammer){
			float Value = sin( fSpeed * (Globals::uRunTimeSeconds + Sync));
			if(Pause > 0.0f){
				const float WP = Width + Pause;
				Value *= (WP);
				Value += Pause;
				if(Value > Width){
					Value = Width;
				}
			}else{
				Value *= Width;
			}
					Value += OriginY;
					y = Value;
					x = OriginX;
		}else{
			Value = ((2*asin(sin(2.0f * fSpeed * (Globals::uRunTimeSeconds + Sync) * 3.14f)))/3.14f);
			if(Pause > 0.0f){
				Value *= (Width + Pause);
				if(Value > Width){
					Value = Width;
				}else if(Value < -Width){
					Value = -Width;
				}
			}else{
				Value *= Width;
			}
			switch(uMoverType){
				case Mover::HorizonalMover:{
					Value += OriginX;
					x = Value;
					y = OriginY;
					break;
				}
				case Mover::VerticalMover:{
					Value += OriginY;
					y = Value;
					x = OriginX;
					break;
				}
				case Mover::Fader:{

					break;
				}
			}
		}
	return;
}
void Mover::SetSpeed(float fNewSpeed){
	fSpeed = fNewSpeed;
	return;
}
void Mover::SetWidth(float fNewMax){
	Width = fNewMax;
	return;
}
void Mover::SetPause(float fNewPause){
	Pause = fNewPause;
	return;
}
float Mover::GetPause(){
	return Pause;
}
float Mover::GetSpeed(){
	return fSpeed;
}
float Mover::GetWidth(){
	return Width;
}

void Mover::SetFunctionType(unsigned int uNewType){
	uFunctionType = uNewType;
	return;
}
void Mover::SetMoverType(unsigned int uNewType){
	uMoverType = uNewType;
	return;
}

unsigned char Mover::GetFunctionType(){
	return uFunctionType;
}
unsigned int Mover::GetMoverType(){
	return uMoverType;
}
